File: algorithms/binarySearch.go
package algorithms

import (
	"math"
	"sort"
)

func BinarySearch(arr []int, value int) int {
	sort.Slice(arr, func(i, j int) bool {
		return arr[i] < arr[j]
	})
	return binarySearchRec(arr, value, 0, len(arr))
}

func binarySearchRec(arr []int, value, low, high int) int {
	middle := int(math.Floor(float64(low + (high-low)/2)))

	if arr[middle] == value {
		return middle
	}

	if value > arr[middle] {
		return binarySearchRec(arr, value, middle+1, high)
	}
	if value < arr[middle] {
		return binarySearchRec(arr, value, low, middle)
	}

	return -1
}

File: grokking_algorithms/binary_search.py
import math


def binary_search(input, target):
    low = 0
    high = len(input) - 1

    while low <= high:
        mid = math.floor((high + low) / 2)

        if input[mid] == target:
            return mid

        if input[mid] > target:
            high = mid - 1
        else:
            low = mid + 1

    return -1


my_list = [1, 3, 5, 7, 9]

print(binary_search(my_list, 5))
print(binary_search(my_list, -1))
